crates/ruff_linter/src/rules/ruff/rules/missing_fstring_syntax.rs

```diff
@@ -0,0 +1,103 @@
+use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{
+    Arguments, Expr, ExprCall, Keyword, Stmt, StmtExpr, StringLiteral, StringLiteralValue,
+};
+use ruff_python_semantic::SemanticModel;
+use ruff_source_file::Locator;
+
+use super::helpers::should_be_fstring;
+
+/// # What it does
+/// Checks for strings that contain f-string syntax but are not f-strings.
+///
+/// ## Why is this bad?
+/// An f-string missing an `f` at the beginning won't format anything, and instead
+/// treat the interpolation syntax as literal.
+///
+/// ## Example
+///
+/// ```python
+/// name = "Sarah"
+/// dayofweek = "Tuesday"
+/// msg = "Hello {name}! It is {dayofweek} today!"
+/// ```
+///
+/// It should instead be:
+/// ```python
+/// name = "Sarah"
+/// dayofweek = "Tuesday"
+/// msg = f"Hello {name}! It is {dayofweek} today!"
+/// ```
+#[violation]
+pub struct MissingFStringSyntax {
+    literal: String,
+}
+
+impl AlwaysFixableViolation for MissingFStringSyntax {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self { literal } = &self;
+        format!(r#"`{literal}` may be a formatting string without an `f` prefix."#)
+    }
+
+    fn fix_title(&self) -> String {
+        "Add an `f` prefix to the formatting string".into()
+    }
+}
+
+pub(crate) fn missing_fstring_syntax(
+    diagnostics: &mut Vec<Diagnostic>,
+    value: &StringLiteralValue,
+    locator: &Locator,
+    semantic: &SemanticModel,
+) {
+    if let Stmt::Expr(StmtExpr { value, .. }) = semantic.current_statement() {
+        if let Expr::StringLiteral(_) = value.as_ref() {
+            return;
+        }
+    }
+
+    let kwargs = get_func_keywords(semantic);
+
+    for literal in value.as_slice() {
+        if should_be_fstring(&literal.value, kwargs, semantic) {
+            let mut diagnostic = Diagnostic::new(
+                MissingFStringSyntax {
+                    literal: locator.slice(literal.range).to_string(),
+                },
+                literal.range,
+            );
+            diagnostic.set_fix(fix_fstring_syntax(literal, locator));
+            diagnostics.push(diagnostic);
+        }
+    }
+}
+
+fn get_func_keywords<'a>(semantic: &'a SemanticModel) -> Option<&'a Vec<Keyword>> {
+    for expr in [
+        semantic.current_expression_parent(),
+        semantic.current_expression_grandparent(),
+    ]
```

Comment:
Should we iterate over _all_ parent expressions here? Or is it intentionally limited to the next two?

Reply:
I want to understand the same. Is it possible that we capture unwanted keyword arguments (one test case suggests otherwise) or that we don't capture the necessary keyword arguments? An example would be helpful as well to better understand the reasoning behind this.

Reply:
This was intentional, though it's not clarified very well in the code itself.

>  Is it possible that we capture unwanted keyword arguments (one test case suggests otherwise) or that we don't capture the necessary keyword arguments?

The idea was to handle cases like `"{a}".format(a = 4)` where the parent expression is `Expr::Attribute` instead of `Expr::Call` - we have to go up one more layer for that.

The reason we only check the first two parents for a call with keyword arguments is because I didn't see a reason to go any higher.